File: asterix-rs/examples/incremental_parsing.rs
//! Example: Incremental parsing of large ASTERIX files
//!
//! This example demonstrates how to use parse_with_offset for processing
//! large ASTERIX data files or streams incrementally.
//!
//! Usage:
//!   cargo run --example incremental_parsing -- <file.asterix> [blocks_per_chunk]

use asterix::{init_default, parse_with_offset, ParseOptions};
use std::env;
use std::fs;
use std::process;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    env_logger::init();

    // Parse command line
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <asterix_file> [blocks_per_chunk]", args[0]);
        eprintln!("\nExample:");
        eprintln!("  {} large_file.pcap 100", args[0]);
        process::exit(1);
    }

    let filename = &args[1];
    let blocks_per_chunk = if args.len() >= 3 {
        args[2].parse().unwrap_or(100)
    } else {
        100
    };

    // Initialize
    println!("Initializing ASTERIX decoder...");
    init_default()?;

    // Read file
    println!("Reading file: {}", filename);
    let data = fs::read(filename)?;
    println!("File size: {} bytes", data.len());

    // Parse incrementally
    let options = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: None,
    };

    let mut offset = 0;
    let mut total_records = 0;
    let mut chunk_num = 0;

    println!("\nParsing in chunks of {} blocks...\n", blocks_per_chunk);

    loop {
        chunk_num += 1;
        print!("Chunk {}: ", chunk_num);

        let result = match parse_with_offset(&data, offset, blocks_per_chunk, options.clone()) {
            Ok(r) => r,
            Err(e) => {
                eprintln!("\nParse error at offset {}: {}", offset, e);
                break;
            }
        };

        let chunk_records = result.records.len();
        total_records += chunk_records;

        println!(
            "parsed {} records, consumed {} bytes, {} blocks remaining",
            chunk_records,
            result.bytes_consumed - offset,
            result.remaining_blocks
        );

        offset = result.bytes_consumed;

        // Check if we're done
        if result.remaining_blocks == 0 || chunk_records == 0 {
            println!("\nReached end of data");
            break;
        }

        // Safety check to prevent infinite loop
        if offset >= data.len() {
            println!("\nReached end of file");
            break;
        }
    }

    println!("\nParsing complete!");
    println!("Total records parsed: {}", total_records);
    println!("Total bytes processed: {}/{}", offset, data.len());

    Ok(())
}

File: asterix-node/examples/incremental_parsing.js
/**
 * Incremental Parsing Example
 *
 * This example demonstrates incremental parsing for large ASTERIX files
 * that don't fit in memory, or for processing data in chunks.
 *
 * Incremental parsing is useful for:
 * - Large files (> 1GB)
 * - Streaming data from network
 * - Memory-constrained environments
 * - Real-time processing with progress reporting
 */

const asterix = require('asterix-decoder');
const fs = require('fs');
const path = require('path');

// Initialize parser
console.log('Initializing ASTERIX decoder...');
asterix.init();

// Example file path
const dataPath = path.join(__dirname, '..', '..', 'install', 'share', 'asterix', 'samples', 'cat_062_065.pcap');

console.log(`\nIncremental parsing: ${dataPath}`);

// Check if file exists
if (!fs.existsSync(dataPath)) {
    console.error(`File not found: ${dataPath}`);
    console.log('\nNote: Build and install the C++ executable first:');
    console.log('  cmake -B build -DCMAKE_BUILD_TYPE=Release');
    console.log('  cmake --build build --parallel');
    console.log('  cmake --install build');
    process.exit(1);
}

try {
    // Read entire file
    const allData = fs.readFileSync(dataPath);
    console.log(`File size: ${allData.length} bytes (${(allData.length / 1024).toFixed(2)} KB)`);

    // Configuration
    const CHUNK_SIZE = 100; // Parse 100 data blocks at a time
    const PROGRESS_INTERVAL = 500; // Report progress every 500 records

    // Statistics
    let totalRecords = 0;
    let offset = 0;
    let iteration = 0;
    const startTime = Date.now();
    const categoryCount = {};
    const itemCount = {};

    console.log(`\nParsing in chunks of ${CHUNK_SIZE} blocks...`);
    console.log('Progress: |' + '='.repeat(50) + '|');
    console.log('          |', { end: '' });

    let lastProgressChars = 0;

    // Incremental parsing loop
    while (offset < allData.length) {
        iteration++;

        // Parse next chunk
        const result = asterix.parseWithOffset(allData, offset, CHUNK_SIZE, {
            verbose: false // Disable verbose for performance
        });

        // Update offset
        offset = result.bytesConsumed;

        // Process records from this chunk
        for (const record of result.records) {
            totalRecords++;

            // Count by category
            categoryCount[record.category] = (categoryCount[record.category] || 0) + 1;

            // Count data items
            for (const itemId of Object.keys(record.items)) {
                itemCount[itemId] = (itemCount[itemId] || 0) + 1;
            }

            // Progress reporting
            if (totalRecords % PROGRESS_INTERVAL === 0) {
                const progressChars = Math.floor((offset / allData.length) * 50);
                if (progressChars > lastProgressChars) {
                    process.stdout.write('='.repeat(progressChars - lastProgressChars));
                    lastProgressChars = progressChars;
                }
            }
        }

        // Check if done
        if (result.remainingBlocks === 0) {
            console.log('='.repeat(50 - lastProgressChars) + '|');
            break;
        }

        // Safety check to prevent infinite loops
        if (iteration > 10000) {
            console.error('\nToo many iterations - possible infinite loop');
            break;
        }
    }

    const elapsedTime = Date.now() - startTime;

    // Display results
    console.log(`\n\n=== Incremental Parsing Complete ===`);
    console.log(`Total records: ${totalRecords}`);
    console.log(`Iterations: ${iteration}`);
    console.log(`Average chunk size: ${(totalRecords / iteration).toFixed(2)} records`);
    console.log(`Parse time: ${elapsedTime}ms`);
    console.log(`Throughput: ${(totalRecords / elapsedTime * 1000).toFixed(0)} records/sec`);
    console.log(`Bytes processed: ${offset} / ${allData.length}`);

    // Category distribution
    console.log(`\n=== Records by Category ===`);
    const sortedCategories = Object.entries(categoryCount).sort((a, b) => Number(a[0]) - Number(b[0]));
    for (const [cat, count] of sortedCategories) {
        const catNum = Number(cat);
        const desc = asterix.isCategoryDefined(catNum) ? asterix.describe(catNum) : 'Unknown';
        const percentage = (count / totalRecords * 100).toFixed(1);
        console.log(`  CAT${cat.padStart(3, '0')}: ${count.toString().padStart(5)} (${percentage}%) - ${desc}`);
    }

    // Most common data items
    console.log(`\n=== Most Common Data Items ===`);
    const sortedItems = Object.entries(itemCount).sort((a, b) => b[1] - a[1]);
    for (const [itemId, count] of sortedItems.slice(0, 10)) {
        const percentage = (count / totalRecords * 100).toFixed(1);
        console.log(`  ${itemId}: ${count.toString().padStart(5)} (${percentage}%)`);
    }

    // Memory efficiency comparison
    console.log(`\n=== Memory Efficiency ===`);
    console.log(`Full parse would require: ${(allData.length / 1024 / 1024).toFixed(2)} MB`);
    console.log(`Incremental parsing chunk: ~${(CHUNK_SIZE * 100 / 1024).toFixed(2)} KB estimated`);
    console.log(`Memory savings: ~${(100 - (CHUNK_SIZE * 100 / allData.length * 100)).toFixed(1)}%`);

    // Example: Filter records by category during incremental parsing
    console.log(`\n=== Filtered Incremental Parsing (CAT062 only) ===`);
    offset = 0;
    let cat062Count = 0;
    const cat062StartTime = Date.now();

    while (offset < allData.length) {
        const result = asterix.parseWithOffset(allData, offset, CHUNK_SIZE, {
            filterCategory: 62,
            verbose: false
        });

        cat062Count += result.records.length;
        offset = result.bytesConsumed;

        if (result.remainingBlocks === 0) {
            break;
        }
    }

    const cat062ParseTime = Date.now() - cat062StartTime;
    console.log(`CAT062 records: ${cat062Count}`);
    console.log(`Parse time: ${cat062ParseTime}ms`);
    console.log(`Speedup vs full parse: ${(elapsedTime / cat062ParseTime).toFixed(2)}x`);

    // Export summary
    const summaryPath = path.join(__dirname, '..', 'incremental_summary.json');
    const summary = {
        metadata: {
            source: dataPath,
            parsed_at: new Date().toISOString(),
            chunk_size: CHUNK_SIZE,
            total_records: totalRecords,
            total_bytes: offset,
            parse_time_ms: elapsedTime,
            iterations: iteration
        },
        categories: categoryCount,
        top_items: Object.fromEntries(sortedItems.slice(0, 20))
    };

    fs.writeFileSync(summaryPath, JSON.stringify(summary, null, 2));
    console.log(`\nSummary exported to: ${summaryPath}`);

} catch (err) {
    console.error('\nError during incremental parsing:', err.message);
    console.error('Stack:', err.stack);
    process.exit(1);
}

console.log('\n=== Example complete ===');
